Implement unit tests for Protocols/CRCElements (#281)
pez-globo/pufferfish-vent-software: firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

/* compute_body_crc

> Scenario: CRCElement: The compute_body_crc function correctly computes crc from body

Reviewer: ```suggestion
Scenario: CRCElement: The compute_body_crc method correctly computes the CRC of a payload from a CRCElement body given as input
```

  GIVEN("A CRC element initialized with an empty payload")

>     WHEN("the crc is computed from the payload")

Reviewer: ```suggestion
    WHEN("the CRC is computed on a body with an empty payload")
```
`compute_body_crc` ignores the payload which the CRCElement is initialized with. All WHENs in this scenario should be reworded in a similar way, to be specific that the payload which compute_body_crc cares about is in the input buffer, not in a CRCElement object.

Author: using this https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r581883306
will change if necessary.

>       THEN("the computed crc is equal to the expected crc")

Reviewer: ```suggestion
      THEN("the computed CRC is 0x00000000")
```
In all THENs in this scenario, we should explicitly list the expected CRC.

>   GIVEN("A CRC Element initialized with the payload '123456789'")

Reviewer: All tests should work on CRCElements initialized with an empty payload, so all GIVENs in this scenario should be GIVEN("A CRC element initialized with an empty payload").

Author: Fixed.

>     WHEN("the crc is computed from the payload")

Author: ```suggestion
    WHEN("the crc is computed on an input_buffer with payload '123456789'")
```
Is this better?

Reviewer: We'll say "an input_buffer body", but it looks good

      THEN("the computed crc is equal to the expected crc")

  GIVEN("A crc element is initalised with payload 0x00")
    WHEN("the crc is computed from the payload")
      THEN("the computed crc is equal to the expected crc")

  GIVEN("A crc element is initalised with payload 0x01")
    WHEN("the crc is computed from the payload")
      THEN("the computed crc is equal to the expected crc")

/* write function

> Scenario: CRCElement: The write function correctly generates body from computed crc and payload

Reviewer: ```suggestion
Scenario: ConstructedCRCElement: The write method correctly generates a body from the payload given in the constructor
```
Part of generating the body is computing a CRC on the payload, so we don't need to state it explicitly in the scenario description.
Should we test the `write` method using a `ConstructedCRCElement`? This way we wouldn't need to add THENs asserting that the payload given in the constructor remains unchanged after a `write`, because the compiler enforces constness.

Author: > Should we test the write method using a ConstructedCRCElement? This way we wouldn't need to add THENs asserting that the payload given in the constructor remains unchanged after a write, because the compiler enforces constness.

Yes, updated the tests for  `write` by using ConstructedCRCElement.

  GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]")

>     WHEN("The crc and payload are written to the output buffer")

Reviewer: Add: THEN("the crc accessor method returns 0 before the write method is called")

Author: Added.

>       THEN("crc computed in write method is same as crcchecker compute")

Reviewer: ```suggestion
      THEN("After the write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
```
This also applies to other tests in this scenario.

>       THEN("The output buffer is as expected")

Reviewer: Add a test above this asserting that the header's CRC field has the correct CRC in the correct position
THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")

Also add a test above this asserting that the body has the correct payload in the correct position:
THEN("The body's payload section correctly stores the payload as 0x01 0x02 0x05.")

This also applies to other tests in this scenario.

Author: Fixed.

  GIVEN("A crc element is initalised with an empty payload")
    WHEN("The crc and payload are written to the output buffer")
      THEN("The CRC returned by the crc() method is equal to 0")
      THEN("The output buffer is as expected")

/* parse function

> Scenario: CRCElement: The parse function correctly updates internal crc and payload fields from the input buffer

Reviewer: ```suggestion
Scenario: CRCElement: The parse method correctly updates its internal crc value and the constructor's payload buffer from the input buffer
```
The way we retrieve the parsed payload is by accessing the payload buffer we gave to the constructor, so this suggested description doesn't imply that that buffer is internal.

>   GIVEN("A CRC Element initialized with an empty payload buffer with a capacity of 254 bytes")

Reviewer: ```suggestion
  GIVEN("A CRCElement constructed with an empty payload buffer with a capacity of 254 bytes")
```

>     WHEN("A body with crc more than 4 bytes and equal to the crc of the payload is parsed")

Reviewer: ```suggestion
    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is parsed")
```
I don't understand what you're saying, the suggested change is my best interpretation but I might've misinterpreted this.

Author: ya my bad , I guess I meant to write a 4 byte CRC , you have rightly interpreted it.

      THEN("the parse method reports ok status")

>       THEN("crc returned from the crc() method is same as the crc of the payload in input body")

Reviewer: ```suggestion
      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
```
This assertion should always be true, whether or not the payload is consistent with the CRC. So we should assert it more directly in this WHEN as well as the other WHENs, rather than relying on the fact that they're all equal in this WHEN.

>       THEN("payload returned from the payload() method is equal to the payload from the body")

Reviewer: ```suggestion
      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
```
Add: THEN("the payload given in the CRCElement constructor is independent of the input body"), i.e. when we edit the data inside the input buffer used for the parse method, the data inside the payload buffer does not change.

Add: THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor") - i.e. when we edit the data in the payload buffer given to the constructor, then the data will also be edited in whatever the `payload()` accessor method returns

This also applies to line 48.

>     WHEN("A body of less than 4 bytes is parsed")

Reviewer: ```suggestion
    WHEN("A body without a complete 4-byte header is parsed")
```

      THEN("the parse method reports out of bounds status")

>     WHEN("A body with crc and payload is given, where the crc is not equal to the actual crc of the payload")

Reviewer: ```suggestion
    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is parsed")
```

      THEN("the parse method reports ok status")

>       THEN("The crc returned from the crc() method is equal to the crc in the body/input_buffer")

Reviewer: ```suggestion
      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
```
This suggested THEN was copied directly from my suggestion on line 38.

>       THEN("The crc returned from the crc() method is not equal to the actual crc for the given payload")

Reviewer: ```suggestion
      THEN("the value returned by the crc accessor method is inconsistent with the payload")
```

      THEN("The payload returned from the payload() method is equal to the payload from the body")

/* write parse roundtrip

> Scenario: The crc element correctly writes and parses to a buffer

Reviewer: ```suggestion
Scenario: CRCElement correctly preserves data in roundtrip writing and parsing
```

  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")

    WHEN("A body with crc and payload is generated, parsed and generated again")
      # write  

>       THEN("The status of write function returns ok")

Reviewer: ```suggestion
      THEN("The first write method call returns ok status")
```
We should specify first vs. second for the THENs asserting things about the write method calls.

      THEN("The crc from the crc element is equal to expected")
      THEN("The payload of the crc element is as expected")
      THEN("The output buffer is as expected")
      # parse
      THEN("The status of parse function returns ok")
      THEN("The crc from the CRCElement is equal to the crc in the body")
      THEN("payload from the CRCElement is equal to the expected payload from the body")
      # write  
      THEN("The status of write function returns ok")
      THEN("The crc from the crc element is equal to expected")

>       THEN("The payload of the crc element is as expected")

Reviewer: Comments and suggestions made about THENs for the individual write and parse scenarios above also apply here. In general these THENs should be changed to look as similar as possible to those previous THENs.

      THEN("The output buffer is as expected")

/* CRCElementReceiver

> Scenario: CRCElementReceiver: The crc element receiver correctly parses datagrams into payloads

Reviewer: ```suggestion
Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them
```

Would you also mind fixing https://github.com/pez-globo/pufferfish-software/blob/eb742d07167caf2d1700923b2b94f5fedc3f71d2/firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Protocols/CRCElements.h#L85 by renaming the `output_datagram` parameter to `output_crcelement`? It looks like I had also missed a spot when I was copy-pasting things :)

Author: yep!
also 
https://github.com/pez-globo/pufferfish-software/blob/eb742d07167caf2d1700923b2b94f5fedc3f71d2/firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Protocols/CRCElements.tpp#L79

and

https://github.com/pez-globo/pufferfish-software/blob/eb742d07167caf2d1700923b2b94f5fedc3f71d2/firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Protocols/CRCElements.tpp#L99

maybe datagram to crcelement ?

  GIVEN("A CRC element receiver of capacity 254 bytes")
    WHEN("A body with crc more than 4 bytes and equal to the crc of the payload is given to the crc_element receiver")
      THEN("the transform method reports ok status")
      THEN("crc returned from the crc() method is equal to the crc in the body/input_buffer")
      THEN("payload returned from the payload() method is equal to the payload from the body")

    WHEN("body with 0 bytes is given to the crc element receiver")
      THEN("the transform status is equal to invalid parse")
      THEN("the crc in output crc_element is equal to the given crc in input_buffer")

    WHEN("body with less than 4 bytes is given to the crc element receiver")
      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
      THEN("the transform status is equal to invalid parse") // for each time
      THEN("The crc returned from the crc() method is equal to 0") // for each time

    WHEN("The CRC stored in the input body is inconsistent with the payload stored in the input body")
      THEN("the transform method reports invalid crc status")

>       THEN("the crc returned from the crc() method is equal to the invalid crc given in input body")

Reviewer: These WHENs and THENs should be updated according to the suggestions made on the write scenario WHENs and THENs above.

/* CRCElementSender

> Scenario: CRCElementSender: The crc element receiver correctly generates datagrams from payload

Reviewer: ```suggestion
Scenario: CRCElementSender correctly generates CRCElement bodies from payloads
```

  GIVEN("A CRC element sender of capacity 254 bytes")
    WHEN("A payload is given along with a sufficiently large output buffer")
      THEN("the transform status is ok")

>       THEN("output buffer is as expected, equal to (crc + input_buffer)")

Reviewer: It would be good to add that the CRC and payload are consistent in either this THEN or a new THEN

Author: Added  THEN("The CRC field of the output buffer's header is consistent with the input_payload.")

  GIVEN("A CRC element sender of buffer size 20 bytes")
    WHEN("The output buffer is not large enough for the input payload")

>       THEN("the transform status is equal to invalid_length")

Reviewer: Is it possible to write a similar test for `CRCElementReceiver`? i.e. where the input buffer is too large for body_max_size

Author: Yes I've written a test for this, so should I then keep it in this PR and fix along with issue 282?

Author: Added a temporary change  for this issue in `CRCElementReceiver ` where I'm just returning an `invalid_parse ` 
if `input_buffer ` is greater than `body_max_size `, 

And for the write method I think `CRCElement::write_protected` does a resize instead to handle this issue.

Then maybe we should follow the same approach for `parse ` as well? 
If it's better to change `Vector::copy_from ` to return error codes then should it be in this PR?

Author: current change: 7504e9a1da4104363ef9eb80fa03677a6af17afc

Reviewer: We'll include fixes for CRCElement and Datagram in this PR, and we'll reserve changes to `Vector` for a future PR

Author: Done with this change https://github.com/pez-globo/pufferfish-software/pull/281/commits/f947b80f5794f159f569382d81704313f06492e3

/* compute_body_crc
Scenario: CRCElement: The compute_body_crc method correctly computes the CRC of a payload from a CRCElement body given as input
  GIVEN("A CRC element initialized with an empty payload of capacity 250 bytes")
    WHEN("the crc is computed on an input_buffer with empty payload")
      THEN("the computed CRC is '0x00000000'")

    WHEN("the crc is computed on an input_buffer with payload '123456789'")
      THEN("the computed crc is '0xe3069283'")

    WHEN("the CRC is computed on an input_buffer with payload '0x00' ")
      THEN("the computed crc is '0x527d5351'")

    WHEN("the crc is computed on an input_buffer with payload '0x01' ")
      THEN("the computed crc is '0xa016d052'")

/* write function
Scenario: ConstructedCRCElement: The write method correctly generates a body from the payload given in the constructor

>   GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]")

Reviewer: I mentioned this in a previous comment but forgot to check whether it applied to other tests, but it does: let's use "constructed" instead of "initialized" when we refer to inputs given into the constructor. There might rarely be some tests of stateful classes where we use both, e.g. "GIVEN a StateSynchronizer constructed with ____ schedules/etc. and initialized to be on the 1st schedule entry"

Author: Fixed.

    WHEN("The crc and payload are written to the output buffer")
      THEN("the crc accessor method returns 0 before the write method is called")
      THEN("After the write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
      THEN("The body's payload section correctly stores the payload as 0x01 0x02 0x05")
      THEN("The output buffer is as expected")

  GIVEN("A crc element is initalised with an empty payload")
    WHEN("The crc and payload are written to the output buffer")

>       THEN("the crc accessor method returns 0 before the write method is called")

Reviewer: I think we need a `THEN("After the write method is called, the crc accessor method returns...")` - I think it returns 0?

Author: yes returns 0, added.

      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
      THEN("The body's payload section is empty")
      THEN("The output buffer is as expected")

/* parse function
Scenario: CRCElement: The parse method correctly updates its internal crc value and the constructor's payload buffer from the input buffer
  GIVEN("A CRCElement constructed with an empty payload buffer with a capacity of 254 bytes")

    WHEN("A body without a complete 4-byte header is parsed")

>       THEN("the parse method reports out of bounds status")

Reviewer: I think we should also add a `THEN("the constructor's payload buffer remains unchanged")`

Author: Added.

Reviewer: It looks like this change got reverted somehow?

    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is parsed")
      THEN("the parse method reports ok status")
      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
      THEN("the value returned by the crc accessor method is inconsistent with the payload")
      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
      THEN("the payload given in the CRCElement constructor is independent of the input body")
      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")

  GIVEN("A CRCElement constructed with a non-empty payload buffer of capacity 254 bytes")
    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is parsed")
      THEN("the crc accessor method returns 0 before the parse method is called")
      THEN("the payload returned from the payload accessor method is equal to the payload buffer given in the CRCElement constructor")
      THEN("the parse method reports ok status")

>       THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")

Reviewer: This THEN should be specified as applying after the parse method is called.

Author: Fixed across test cases. also updated for receiver as `After the transform method is called`

      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
      THEN("the payload given in the CRCElement constructor is independent of the input body")
      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")

/* write parse roundtrip
Scenario: CRCElement correctly preserves data in roundtrip writing and parsing

  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")

>     WHEN("A body with crc and payload is generated, parsed and generated again")

Reviewer: I thought about this a bit more and I think the test would be stronger if we didn't reuse the same CRCElement object (and the same payload buffer) for all three write/parse/write method calls, but instead instantiated a new ParsedCRCElement/ConstructedCRCElement object (and a new payload buffer, for ParsedCRCElement objects) for each method call. This way we can ensure that roundtripping works correctly even if no internal state is preserved across those method calls, which is closer to how the objects would be instantiated and used in the application.

Then we could update the WHEN as: WHEN A body with CRC and payload is generated, parsed into a new ParsedCRCElement object, and generated again with a new ConstructedCRCElement object

Author: Updated using ParsedCRCElement & ConstructedCRCElement, using different payload buffers for constructing the elements and different buffers for input.

      # write  
      THEN("The second write method call returns ok status")
      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
      THEN("The output buffer is as expected")

/* CRCElementReceiver
Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them

  GIVEN("A CRC element receiver of capacity 254 bytes")
    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
      THEN("the transform method reports ok status")
      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
      THEN("the payload buffer updated in the CRCElement is independent of the input body")

>     WHEN("A body without a complete 4-byte header and empty payload is given to the crc element receiver")

Reviewer: I don't understand how this WHEN is different from the next WHEN - I think one of these was the original one and the other was added in response to a suggested revision. I would go with something like "WHEN a body with an incomplete header and an empty payload is..." or "WHEN A body without a complete 4-byte header is..." (the latter is a direct copy from tests above, but maybe the tests above should be changed to be more like the former). The current wording, "WHEN A body without a complete 4-byte header and empty payload is...", is a bit ambiguous and could be misinterpreted as meaning "WHEN A body with an incomplete 4-byte header and a non-empty payload is..." (this is the `!A && B` vs. `!(A && B)` problem)

Author: Sorry about the ambiguity, so in this test case I'm basically using an empty `input_buffer`, in the latter test we are using a body with crc header less than 4 bytes, so 1,2 and 3 respectively so there it's the case of `incomplete header`, here there's nothing.
so maybe the WHEN should be something like WHEN("An empty input_buffer is given to the crc element receiver")

Reviewer: Yes, I like that - good idea to test an empty input as distinct from a non-empty too-small input

      THEN("the transform status is equal to invalid parse")
      THEN("the value returned by the crc accessor method is equal to 0x00000000r")

>       THEN("The body's payload section is empty")

Reviewer: This doesn't seem correct to me - the body is exactly the contents of input_buffer, which is const. So the payload section of the body is preserved at what it was. Separately, the value returned by the payload accessor method of output_crcelement will be empty if (and only if) output_crcelement had an empty payload before the CRCElementReceiver::transform was called.

Author: Yes i'm not sure how this got here, I'm actually not checking what payload accessor returns for this case as it's empty also the `input_buffer` is empty to begin with.

Secondly in all CRCElementReceiver test cases I'm indeed using an empty payload buffer for the output_crcelement. 
So should I get rid of this THEN completely or should I do something like
THEN("The payload buffer returned by the payload accessor method is empty") ?

Reviewer: My preference would be something like a combination of the following:
- THEN("after the transform method is called, the payload given in the ParsedCRCElement constructor remains unchanged") // i.e. if we inserted some dummy data into the buffer before giving it to the constructor, the data stays there; and if we give make the buffer empty before giving it to the constructor, the buffer stays empty
- THEN("the payload returned from the payload accessor method is equal to the payload from the body")

Author: Added.

    WHEN("body with less than 4 bytes is given to the crc element receiver")
      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
      THEN("the transform status is equal to invalid parse") // for each time

>       THEN("the value returned by the crc accessor method is equal to 0x00000000") // for each time

Reviewer: I thought about this a bit more and this is not strictly correct (also applies to the same THEN in the previous WHEN). Actually the values returned by the crc and payload methods should just be whatever they were before. For example, if we passed in an output_crcelement with a nonzero crc and a payload with some bytes inside, those should be preserved after transform returns invalid_parse. We should cover this in a test, and I think the WHEN will need to specify what (if anything) was the state of output_crcelement before calling transform.

This test should also specify that the crc accessor method is called on the output_crcelement, not on the CRCElementReceiver object.
This test should also have a THEN checking the value returned by the payload accessor method of output_Crcelement.

Author: Correct! so to give some context, for all the tests in CRCElementReceiver I'm using a `output_crcelement` with empty payload given in the constructor, I did that mainly as we are using it as an output parameter and if parse is succesful it will get updated.

> This test should also specify that the crc accessor method is called on the output_crcelement,

will update this for all cases.

So I was thinking for these I'll divide into two GIVEN's , the ones that are already there with
  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with empty payload")
and
  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with payload '0x12 0x13 0x14 0x15 0x16'")

> . For example, if we passed in an output_crcelement with a nonzero crc and a payload with some bytes inside,

Also even with a output_crcelement with some bytes inside, `crc_ ` member parameter is only updated when `write` method is called, so if we want a test that has crcelement with some crc then we must do that.

Reviewer: Yes, I agree with the approach of into two GIVENs - I would just specify that the output_crcelement is constructed with an empty payload _buffer_, since the way we're using it in this test is as a place to dump data rather than as the input data itself. In both GIVENs, the THEN could just say that output_crcelement's payload buffer remains unchanged, and the only difference is in the test code assertions.

For the GIVEN with a nonzero CRC, yes part of the test setup code would involve calling `output_crcelement.write` and throwing away the output buffer, in order to make the `crc_` private member become nonzero, in order to set up the internal state of that object for the test.

Reviewer: If we can establish the following in the CRCElement tests:

THEN the bytes in the constructor's payload buffer look like "..."
THEN the payload accessor method returns the constructor's payload buffer

is equivalent to:

THEN the CRCElement contains the payload "..."

then in the CRCElementReceiver tests we can just assert THEN the CRCElement contains the payload "..."

Reviewer: Pattern to use for all CRCElementReceiver tests (and CRCElement tests in a similar pattern):

```
GIVEN ...
  WHEN ...
    THEN the transform method returns ... status code
    THEN the ParsedCRCElement output parameter has the following properties:
      - Its crc accessor method returns ...
      - The payload given to its constructor is ...
      - Its payload accessor method returns a reference to the payload given in its constructor
      - The data in its payload is independent of the data in input_buffer
```

    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is given to the receiver")
      THEN("the transform method reports invalid crc status")

>       THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")

Reviewer: This test should also have a THEN checking the value returned by the payload accessor method of output_crcelement.

Author: https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r584449591

So when the GIVEN is a empty payload in output_crcelement, I'll say "returns empty payload buffer"
and for a GIVEN with payload in output_crcelement "returns 'payload'"

Reviewer: I think this will be clearest if we split it into two THENs, as in https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r585177976 - so here I think we'd assert that the contents of the payload buffer which had been given to the constructor are set to the bytes from the payload section of the input body, and then we'd assert that the payload accessor method returns a reference to that payload buffer in the constructor

Author: Added in CRCElement following https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r585243812
https://github.com/pez-globo/pufferfish-software/pull/281/files#diff-10ffb7b9ec39ab80c47393699e09e302b3ba3f871bf74fb4f834bf350cc97ef0R65

in Receiver https://github.com/pez-globo/pufferfish-software/pull/281/files#diff-10ffb7b9ec39ab80c47393699e09e302b3ba3f871bf74fb4f834bf350cc97ef0R132

  GIVEN("A CRC element receiver of capacity 10 bytes") {

    WHEN("A body of size 20 bytes with a complete 4-byte header, And a payload of capacity 16 bytes completely filled with data is given to the receiver")

>       THEN("the transform method reports invalid parse status")

Reviewer: This test should also have THENs checking the values returned by the crc and payload accessor methods of output_crcelement.

Author: Added.

      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor") 
      THEN("the payload given in the CRCElement constructor is independent of the input buffer body")

    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is parsed")
      THEN("the parse method reports ok status")
      THEN("After the parse method is called, the value returned by the crc accessor method is equal to the crc field of the input body's header")
      THEN("the value returned by the crc accessor method is inconsistent with the payload")
      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
      THEN("the payload given in the CRCElement constructor is independent of the input buffer body")

  GIVEN("A CRCElement constructed with a non-empty payload buffer of capacity 254 bytes")

    WHEN("A body without a complete 4-byte header is parsed")
      THEN("the parse method reports out of bounds status")

>       THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")

Reviewer: Actually, the CRC remains at whatever value it was given before parse was called.

>       THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")

Reviewer: I think the CRC remains at whatever value it had before parse was called.

>       THEN("The payload buffer given to it's constructor is '0x12 0x13 0x14 0x15 0x16' ")

Reviewer: I think it would be clearer if this were `THEN("The payload buffer given to its constructor remains unchanged")`

      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
      THEN("the payload given in the CRCElement constructor is independent of the input buffer body")      
  
    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is parsed")
      THEN("the crc accessor method returns 0 before the parse method is called")
      THEN("The bytes in the constructor's payload buffer are '0x12 0x13 0x14 0x15 0x16' ")

>       THEN("The payload accesor method returns the constructor's payload buffer")

Reviewer: I think either these three THENs should also be in the previous WHEN for consistency, or they should be omitted because they can just be REQUIREs under the WHEN, but not explicitly THENs.

      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
      THEN("the payload given in the ParsedCRCElement constructor is independent of the input body")
      # write  
      THEN("The second write method call returns ok status")
      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
      THEN("The output buffer is as expected")

/* CRCElementReceiver
Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them

  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with an empty payload buffer")

    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
      THEN("the transform method reports ok status")

>       THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")

Reviewer: I didn't mean for this to literally be its own THEN - after all, there can't be any REQUIREs associated specifically with it. Rather, it was meant so that the following THENs would substitute "its" with "output_crcelement's":
```
      THEN("the value returned by output_crcelement's crc accessor method is equal to the crc field of the input body's header")
      THEN("the payload given to output_crcelement's constructor is empty")
      THEN("output_crcelement's payload accessor method returns a reference to the payload given in its constructor")
      THEN("The data in output_crcelement's payload is independent of the data in input_buffer")
```
(this also applies to the other tests in this file)

      THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")

>       THEN("the payload given to it's constructor is empty")

Reviewer: The CRCElementReceiver successfully the payload from the body. Is the payload supposed to still be empty?

      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
      THEN("The data in its payload is independent of the data in input_buffer")

    WHEN("An empty input_buffer is given to the crc element receiver")
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")

>       THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")

Reviewer: The input body is empty, so this THEN doesn't make sense to me

      THEN("the payload given to it's constructor is empty")
      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
      THEN("The data in its payload is independent of the data in input_buffer")

    WHEN("body with less than 4 bytes is given to the crc element receiver")
      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")

>       THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")

Reviewer: I'm pretty sure the CRC accessor method returns the same value as whatever it would've returned before calling the transform method, because the parse returned early with out_of_bounds.

      THEN("the payload given to it's constructor is empty")
      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
      THEN("The data in its payload is independent of the data in input_buffer")

    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is given to the receiver")
      THEN("the transform method reports invalid crc status")
      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
      THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")

>       THEN("the payload given to it's constructor is empty")

Reviewer: The CRCElementReceiver successfully the payload from the body. Is the payload supposed to still be empty?

      THEN("The data in its payload is independent of the data in input_buffer")

  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with an non-empty payload buffer")

    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
      THEN("the transform method reports ok status")
      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
      THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")
      THEN("the payload given to it's constructor is '0x01 0x05 0x12 0x13 0x14 0x15 0x16'")
      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
      THEN("The data in its payload is independent of the data in input_buffer")
  
    WHEN("An empty input_buffer is given to the crc element receiver")
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")

>       THEN("it's crc accessor method returns 0")

Reviewer: This THEN would be clearer if you specified in the WHEN what the CRC member was before it was given to the CRCElementReceiver (you do this in the next WHEN)

>       THEN("the payload given to it's constructor is '0x01 0x05 0x12 0x13 0x14 0x15 0x16'")

Reviewer: I think it would be clearer This THEN would be clearer if you specified in the WHEN what the payload was before it was given to the CRCElementReceiver (I would suggest doing this in the next WHEN too)

      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
      THEN("The data in its payload is independent of the data in input_buffer")

>     WHEN("An empty input_buffer is given to the crc element receiver and output_crcelement with a non-zero crc member")

Reviewer: Did you mean: WHEN("An empty input_buffer and an output_crcelement with a non-zero crc member are given to the crc element receiver")

      THEN("The write method of output_crcelement reports ok status")
      THEN("The value returned by the crc accessor method of output_crcelement is '0x40D06D79'")

>       THEN("the transform status is equal to invalid parse")

Reviewer: My comment on lines 73-75 applies here, too

      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
      THEN("the value returned by it's crc accessor method is equal to '0x40D06D79'")

>       THEN("the payload given to it's constructor is '0x01 0x05 0x12 0x13 0x14 0x15 0x16'")

Reviewer: Are these values coming from the input_buffer after a successful parse, or are they the data which was already in output_crcelement and which was left unchanged after an unsuccessful parse? The CRC value appears to be the same before and after transform, while the payload is left unspecified before transform.

I think it'll be clearer and more maintainable if you use the following structure for THENs in this file (don't copy these literally, adapt the language and assertions to make sense for each specific test):
- THEN the value returned by output_crcelement's crc accessor method remains the same as before the transform method call
- THEN the value returned by output_crcelement's crc accessor method becomes equal to the crc field of input_buffer's header
- THEN the value returned by output_crcelement's payload accessor method remains the same as before the transform method call
- THEN the value returned by output_crcelement's payload accessor method becomes equal to the payload of input_buffer

      THEN("the value returned by the crc accessor method is inconsistent with the payload")
      THEN("the payload returned from the payload accessor method is equal to the payload from the input_buffer body")
      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
      THEN("the payload given in the CRCElement constructor is independent of the input buffer body")

  GIVEN("A CRCElement constructed with a non-empty payload buffer of capacity 254 bytes")
    // assertions carried out in GIVEN:
    // The crc accessor method returns 0
    // The payload buffer returned by the payload accessor method is '0x12 0x13 0x14 0x15 0x16'

    WHEN("A body without a complete 4-byte header is parsed")
      THEN("the parse method reports out of bounds status")
      THEN("After the parse method is called, the value returned by the crc accessor method remains unchanged")
      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")

>       THEN("the payload given in the CRCElement constructor is independent of the input buffer body")      

Reviewer: What should the contents of the payload buffer be after parse method is called? (I think the contents should be unchanged)

      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
      THEN("The output buffer is as expected")

/* CRCElementReceiver
Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them

  GIVEN("A CRC element receiver of capacity 254 bytes and an output_crcelement constructed with an empty payload buffer")
    // assertions carried out in GIVEN:
    // The output_crcelement's crc accessor method returns 0
    // The payload buffer returned by the output_crcelement's payload accessor method is empty

    WHEN("An empty input_buffer is given to the crc element receiver")
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method remains unchanged")

>       THEN("The output_crcelement's payload accessor method returns a reference to the payload given in its constructor")

Reviewer: What should the contents of the payload buffer be after parse method is called? (I think the contents should be unchanged)

      THEN("The data in the output_crcelement's payload buffer is independent of the data in input_buffer")

    WHEN("A body with less than 4 bytes is given to the crc element receiver")
      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method remains unchanged")

>       THEN("The output_crcelement's payload accessor method returns a reference to the payload given in its constructor")

Reviewer: What should the contents of the payload buffer be after parse method is called? (I think the contents should be unchanged)

      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method is equal to the crc field of the input_buffer body's header")
      THEN("the value returned by the crc accessor method is inconsistent with the payload")
      THEN("The payload buffer returned by the output_crcelement's payload accessor method is equal to the payload from the input_buffer body")
      THEN("The data in the output_crcelement's payload buffer is independent of the data in input_buffer")

  GIVEN("A CRC element receiver of capacity 254 bytes and an output_crcelement constructed with an non-empty payload buffer")
    // assertions carried out in GIVEN:
    // The output_crcelement's crc accessor method returns 0
    // The payload buffer returned by the output_crcelement's payload accessor method is '0x1e 0xe2 0x9a 0xf8 0x4f'

    WHEN("An empty input_buffer is given to the crc element receiver")
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method remains unchanged")

>       THEN("The output_crcelement's payload accessor method returns a reference to the payload given in its constructor")

Reviewer: What should the contents of the payload buffer be after parse method is called? (I think the contents should be unchanged)

      THEN("the transform method reports ok status")
      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method is equal to the crc field of the input_buffer body's header")
      THEN("The payload buffer returned by the output_crcelement's payload accessor method is equal to the payload from the input_buffer body")
      THEN("The data in the output_crcelement's payload buffer is independent of the data in input_buffer")

  GIVEN("A CRC element receiver of capacity 254 bytes, and output_crcelement constructed with an non-empty payload buffer and non-zero crc member")
    // assertions carried out in GIVEN:
    // The value returned by the output_crcelement's crc accessor method is '0x40D06D79'
    // The payload buffer returned by the output_crcelement's payload accessor method is '0x23 0x84 0xf5 0x6a 0xac 0xae'

    WHEN("An empty input_buffer is given to the crc element receiver")
      THEN("the transform method reports invalid parse status")
      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method remains unchanged")

>       THEN("The output_crcelement's payload accessor method returns a reference to the payload given in its constructor")

Reviewer: What should the contents of the payload buffer be after parse method is called? (I think the contents should be unchanged)

      THEN("the transform method reports ok status")
      THEN("After the transform method is called, the value returned by the output_crcelement's crc accessor method is equal to the crc field of the input_buffer body's header")
      THEN("The payload buffer returned by the output_crcelement's payload accessor method is equal to the payload from the input_buffer body")
      THEN("The data in the output_crcelement's payload buffer is independent of the data in input_buffer")      

/* CRCElementSender
Scenario: CRCElementSender: correctly generates CRCElement bodies from payloads

  GIVEN("A CRC element sender of capacity 254 bytes")

    WHEN("A payload is given along with a sufficiently large output buffer")
      THEN("the transform method reports ok status")
      THEN("output buffer is as expected, equal to (crc + input_buffer)")
      THEN("The CRC field of the output buffer's header is consistent with the input_payload.")

  GIVEN("A CRC element sender of buffer size 20 bytes")

    WHEN("The output buffer is not large enough for the input payload")

>       THEN("the transform status is equal to invalid_length")

Reviewer: What should the contents of output_buffer be after the transform method is called? (I think the contents should be unchanged, whether output_buffer is empty or not)
